### PR TITLE
add raster stac generator

### PR DESCRIPTION
**Pull request by alisha17**

### CODE REVIEW

File: example/csv/generated/raster/orthomosaic_item/orthomosaic_item.json
Comment (JosephUoA) [inline example/csv/generated/raster/orthomosaic_item/orthomosaic_item.json:1]: There are no `eo` fields in any of these examples. There are also no `raster` fields and `raster` extension is not even listed in the `stac_extensions`

File: stac_generator/raster_data/generator.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                )
+                item.add_asset("data", asset)
+
+                # Add STAC extensions
+                item.stac_extensions = [
+                    "https://stac-extensions.github.io/eo/v1.1.0/schema.json",
+                    "https://stac-extensions.github.io/projection/v1.1.0/schema.json"
+                ]
+
+                # Force cleanup of any erroneous eo:bands in item.properties
```
Comment (JosephUoA) [inline stac_generator/raster_data/generator.py]: Why do you do this?

File: stac_generator/raster_data/generator.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                for idx, band_info in enumerate(source_cfg.bands, start=1):
+                    band = Band.create(
+                        name=f"B{idx}",
+                        common_name=BAND_MAPPING.get(band_info.name.lower(), None),
+                        center_wavelength=float(band_info.wavelength) if band_info.wavelength not in [None, "no band specified"] else None
+                    )
+                    eo_bands.append(band)
+
+                # Ensure eo:bands is managed by EOExtension only
+                eo_ext.bands = eo_bands
```
Comment (JosephUoA) [inline stac_generator/raster_data/generator.py]: Why is this necessary?

File: stac_generator/raster_data/generator.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                    media_type=media_type,
+                    roles=["data"],
+                    title="Raster Data"
+                )
+                item.add_asset("data", asset)
+
+                # Add STAC extensions
+                item.stac_extensions = [
+                    "https://stac-extensions.github.io/eo/v1.1.0/schema.json",
+                    "https://stac-extensions.github.io/projection/v1.1.0/schema.json"
```
Comment (JosephUoA) [inline stac_generator/raster_data/generator.py]: What happened to the `raster` extension?

File: example/csv/generated/vector_polygon/vector_polygon_zip/vector_polygon_zip.json [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  "bbox": [
+    434067.17171849194,
+    6950889.883355473,
+    434208.6646524757,
+    6951062.34869196
+  ],
+  "properties": {
+    "start_datetime": "2024-10-28T01:14:24.531804Z",
+    "end_datetime": "2024-10-28T01:14:24.531804Z",
+    "proj:bbox": [
```
Comment (JosephUoA) [inline example/csv/generated/vector_polygon/vector_polygon_zip/vector_polygon_zip.json:50]: Should `proj:epsg` be in here?

File: example/csv/generated/vector_polygon/vector_polygon_geojson/vector_polygon_geojson.json [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  "bbox": [
+    138.62863776330823,
+    -34.96914105537954,
+    138.6294879005062,
+    -34.96855545871302
+  ],
+  "properties": {
+    "start_datetime": "2024-10-28T01:14:24.531743Z",
+    "end_datetime": "2024-10-28T01:14:24.531743Z",
+    "proj:bbox": [
```
Comment (JosephUoA) [inline example/csv/generated/vector_polygon/vector_polygon_geojson/vector_polygon_geojson.json:45]: Should `proj:epsg` be in here?

**Comment by JosephUoA** [comment]
Please check raster stac examples and make sure missing fields in your examples are added. Also fix the CI.

File: stac_generator/raster_data/generator.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                    "instruments": ["Multispectral Camera"],
+                })
+
+                # Add projection extension
+                proj_ext = ProjectionExtension.ext(item, add_if_missing=True)
+                if source_cfg.epsg != crs.to_epsg():
+                    raise ValueError(f"EPSG mismatch: {source_cfg.epsg} vs {crs.to_epsg()}")
+                proj_ext.epsg = crs.to_epsg()
+                proj_ext.transform = list(transform)[:6]
+                proj_ext.shape = [src.height, src.width]
```
Comment (JosephUoA) [inline stac_generator/raster_data/generator.py]: I think you can just do `proj_ext.shape = src.shape`

File: stac_generator/raster_data/generator.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+                asset = pystac.Asset(
+                    href=str(source_cfg.location),
+                    media_type=media_type,
+                    roles=["data"],
+                    title="Raster Data"
+                )
+                item.add_asset("data", asset)
+
+                # Add STAC extensions
+                item.stac_extensions = [
```
Comment (JosephUoA) [inline stac_generator/raster_data/generator.py]: Its been a while since I wrote the prototype but Im pretty sure this should be populated automatically by pystac. Did you use the existing GeotiffGenerator class as a reference? There are many things that code does, or comments it contains that are not addressed here.

File: example/csv/generated/raster/orthomosaic_item/orthomosaic_item.json
Comment (alisha17) [reply example/csv/generated/raster/orthomosaic_item/orthomosaic_item.json:1]: Yes correctly pointed. Still fixing some issues with the code.

Comment (alisha17) [reply example/csv/generated/raster/orthomosaic_item/orthomosaic_item.json:1]: this is fixed in the latest update

File: stac_generator/raster_data/generator.py (hunk 2 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: Removed in latest update

File: example/csv/generated/vector_polygon/vector_polygon_zip/vector_polygon_zip.json (hunk 4 above)
Comment (alisha17) [reply example/csv/generated/vector_polygon/vector_polygon_zip/vector_polygon_zip.json:50]: Yep - added

File: example/csv/generated/vector_polygon/vector_polygon_geojson/vector_polygon_geojson.json (hunk 5 above)
Comment (alisha17) [reply example/csv/generated/vector_polygon/vector_polygon_geojson/vector_polygon_geojson.json:45]: yep - added

File: stac_generator/raster_data/generator.py (hunk 6 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: removed in latest update

File: stac_generator/raster_data/generator.py (hunk 7 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: correct - removed in latest update

File: stac_generator/raster_data/generator.py [hunk 8]
```diff
@@ -29,105 +39,133 @@ def __init__(
...
+                    ]
+                )
+                geometry_geojson = mapping(geometry)
+
+                if source_cfg.datetime is None:
+                    raise ValueError("source_cfg.datetime cannot be None")
+
+                # Ensure datetime is timezone-aware
+                if source_cfg.datetime.tzinfo is None:
+                    datetime_aware = source_cfg.datetime.replace(tzinfo=UTC)
```
Comment (JosephUoA) [inline stac_generator/raster_data/generator.py]: I think it will be problematic to assume UTC. There is probably not going to be any timezone information in the source_cfg. Timezone is a function of location though and we do have that information. Im not sure of the best way to do this but there are almost certainly packages that give you the timezone from a lat/lon or other location input.

File: stac_generator/raster_data/generator.py [hunk 9]
```diff
@@ -29,105 +39,133 @@ def __init__(
...
+                    )
+                    raster_bands.append(raster_band)
 
-                # Update other properties
-                item.properties.update({
-                    "gsd": float(src.res[0]),
-                    "platform": "UAV",
-                    "instruments": ["Multispectral Camera"],
-                })
+                item.properties["eo:bands"] = [band.to_dict() for band in eo_bands]
```
Comment (JosephUoA) [inline stac_generator/raster_data/generator.py]: Great work getting the band attributes from the config file. On line 152 and 153 you add the raster extension fields using the apply method on the extension but here and on line 156 you are manually adding to keys in a dictionary. I prefer the method you used for the raster extension.

File: stac_generator/raster_data/generator.py [hunk 10]
```diff
@@ -29,105 +39,133 @@ def __init__(
...
-                        [bounds.left, bounds.bottom],
-                        [bounds.left, bounds.top],
-                        [bounds.right, bounds.top],
-                        [bounds.right, bounds.bottom],
-                        [bounds.left, bounds.bottom]
-                    ]]
-                }
-
-                # Create base item
+                shape = [src.height, src.width]
```
Comment (JosephUoA) [inline stac_generator/raster_data/generator.py]: You should be able to get the shape from the geotiff above with `shape = src.shape`. I think this way is preferable because it removes the possibility of getting the ordering wrong here.

File: stac_generator/raster_data/generator.py [hunk 11]
```diff
@@ -29,105 +39,133 @@ def __init__(
...
+                        description=f"Common name: {BAND_MAPPING.get(band_info.name.lower(), 'unknown')}",
                     )
-                    eo_bands.append(band)
+                    eo_bands.append(eo_band)
 
-                # Ensure eo:bands is managed by EOExtension only
-                eo_ext.bands = eo_bands
+                    raster_band = RasterBand.create(
+                        nodata=band_info.nodata or 0,
+                        data_type=DataType(band_info.data_type or "uint16"),
```
Comment (JosephUoA) [inline stac_generator/raster_data/generator.py:161]: The data type doesn't come from the config and cant be set as `uint16` default. It is encoding in the source file and should be read from there instead. I think in this case it would be `src.dtypes`.

File: stac_generator/raster_data/generator.py [hunk 12]
```diff
@@ -29,105 +39,133 @@ def __init__(
...
-                    geometry=geometry,
+                    geometry=geometry_geojson,
                     bbox=bbox,
-                    datetime=source_cfg.datetime,
-                    properties={}  # properties will be managed by extensions
+                    datetime=datetime_aware,
+                    properties={
+                        "eo:snow_cover": 0,
+                        "eo:cloud_cover": 0,
+                        "proj:epsg": crs.to_epsg(),
```
Comment (JosephUoA) [inline stac_generator/raster_data/generator.py]: As I suggested with the vector stac generator, you should compare that the epsg code provided in the config file is the same as the one read from the file.

File: stac_generator/raster_data/schema.py [hunk 13]
```diff
@@ -1,93 +1,99 @@
...
-                "%H:%M:%S%z"
-            ).time()
-            
-            # Combine date and time
-            collection_datetime = datetime.combine(
-                collection_date,
-                collection_time
+            # Parse the date string as timezone-aware
+            collection_date = datetime.strptime(self.collection_date, "%Y-%m-%d").replace(
+                tzinfo=UTC
```
Comment (JosephUoA) [inline stac_generator/raster_data/schema.py]: See comment on stac_generator/raster_data/generator.py regarding time zones.

**Comment by JosephUoA** [comment]
Lets have a call to clarify some of the points I have raised here.

File: stac_generator/raster_data/generator.py (hunk 8 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: So we can get the location from geotiff file. Use a point or centroid.

File: stac_generator/raster_data/generator.py (hunk 9 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: Use .apply()

File: stac_generator/raster_data/generator.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+from pystac.extensions.raster import AssetRasterExtension, DataType, RasterBand
+from shapely.geometry import Polygon, mapping
+
+from stac_generator.base.generator import StacGenerator
+from stac_generator.base.schema import StacCatalogConfig, StacCollectionConfig
+
+from .schema import RasterSourceConfig
+
+# In case the config file has different names for bands
+BAND_MAPPING: dict[str, str] = {
```
Comment (harryle95) [inline stac_generator/raster_data/generator.py:22]: I am not sure what's the purpose of this since it seems to map to the same thing

File: stac_generator/raster_data/schema.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+
+class RasterSourceConfig(SourceConfig):
+    """Configuration for raster data sources"""
+
+    location: str
+    """Location of the raster file"""
+    epsg: int
+    """EPSG code for the raster's coordinate reference system"""
+    collection_date: str  # Changed to str to handle the input format
```
Comment (harryle95) [inline stac_generator/raster_data/schema.py:35]: Not really sure if `collection_date` here means the datetime when data was acquired or the datetime field of the collection. Either way, these should not be used and the fields `datetime`, `start_datetime`, and `end_datetime` native to STAC metadata should be used instead.

File: stac_generator/raster_data/schema.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+            # Ensure each band is properly formatted
+            if isinstance(data.get("bands"), list):
+                data["bands"] = [
+                    {"name": band, "wavelength": None} if isinstance(band, str) else band
+                    for band in data["bands"]
+                ]
+        return data
+
+    @model_validator(mode="after")
+    def set_datetime_from_collection(self) -> RasterSourceConfig:
```
Comment (harryle95) [inline stac_generator/raster_data/schema.py]: This method shouldn't exist. Pydantic should automatically coerce the date fields should they be in the right format - i.e. you don't need to write your own date parser.

File: stac_generator/raster_data/schema.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    """Band information for raster data"""
+
+    name: str
+    wavelength: float | None = Field(default=None)  # Can be float or None
+    nodata: float | None = Field(default=0)  # Default nodata value
+    data_type: str | None = Field(default="uint16")  # Default data type for raster band
+
+    @model_validator(mode="before")
+    @classmethod
+    def parse_wavelength(cls, data: Any) -> Any:
```
Comment (harryle95) [inline stac_generator/raster_data/schema.py:21]: I don't know whether this is necessary. If there is no wavelength information, they should just leave it None or blank instead of "no band specified". This should also just be a TypedDict since it doesn't need validation.

File: stac_generator/raster_data/generator.py [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+        source_df: pd.DataFrame,
+        collection_cfg: StacCollectionConfig,
+        catalog_cfg: StacCatalogConfig | None = None,
+        href: str | None = None,
+    ) -> None:
+        # Reset the DataFrame index
+        source_df = source_df.reset_index(drop=True)
+
+        # Ensure the EPSG column is properly formatted
+        if "epsg" in source_df.columns:
```
Comment (harryle95) [inline stac_generator/raster_data/generator.py]: You shouldn't need to redo type validation. Pydantic should have already handled this

File: stac_generator/raster_data/schema.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+            self.start_datetime = collection_datetime
+            self.end_datetime = collection_datetime
+
+        except ValueError as e:
+            raise ValueError(f"Error parsing date/time: {e}") from e
+
+        return self
+
+    class Config:
+        arbitrary_types_allowed = True
```
Comment (harryle95) [inline stac_generator/raster_data/schema.py]: Don't do this

File: stac_generator/raster_data/generator.py [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+            source_df["epsg"] = source_df["epsg"].astype(int)
+
+        # Parse the 'bands' column into a list if it is a JSON-encoded string
+        if "bands" in source_df.columns:
```
Comment (harryle95) [inline stac_generator/raster_data/generator.py:45]: This should also be handled at the schema level, not here

File: stac_generator/raster_data/schema.py [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+    location: str
+    """Location of the raster file"""
+    epsg: int
+    """EPSG code for the raster's coordinate reference system"""
+    collection_date: str  # Changed to str to handle the input format
+    collection_time: str  # Changed to str to handle the input format
+    bands: list[BandInfo]
+    """List of band information"""
+    transform: list[float] | None = Field(default=None)
```
Comment (harryle95) [inline stac_generator/raster_data/schema.py]: I am not sure if `transform` and `shape` in the config are used anywhere. Was thinking they might be compared against extracted information like epsg but this was not done in the generator code.

File: stac_generator/raster_data/generator.py [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+                item.properties["eo:bands"] = [band.to_dict() for band in eo_bands]
+
+                # Create Asset and Add to Item
+                asset = pystac.Asset(
+                    href=source_cfg.location,
+                    media_type=pystac.MediaType.GEOTIFF,
+                    roles=["data"],
+                    title="Raster Data",
+                )
+                item.add_asset("image", asset)
```
Comment (harryle95) [inline stac_generator/raster_data/generator.py]: For consistency, let's use "data" as asset name since this was used for both the vector and point generator. This is for the stac loader.

File: stac_generator/raster_data/generator.py (hunk 14 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py:22]: Yes, but the name of the band might be a bit different (uppercase, lowercase, space between red and edge for eg.), so its more for the future if we wanted to update

File: stac_generator/raster_data/schema.py (hunk 16 above)
Comment (alisha17) [reply stac_generator/raster_data/schema.py]: removed

File: stac_generator/raster_data/schema.py (hunk 19 above)
Comment (alisha17) [reply stac_generator/raster_data/schema.py]: removed

File: stac_generator/raster_data/generator.py (hunk 22 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: done

File: stac_generator/raster_data/generator.py (hunk 8 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: done

File: stac_generator/raster_data/generator.py (hunk 10 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: done

File: stac_generator/raster_data/generator.py (hunk 12 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: done

File: stac_generator/raster_data/schema.py (hunk 13 above)
Comment (alisha17) [reply stac_generator/raster_data/schema.py]: removed

File: stac_generator/raster_data/generator.py (hunk 9 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: done

File: stac_generator/raster_data/generator.py (hunk 11 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py:161]: using src.dtypes now

File: stac_generator/raster_data/generator.py (hunk 14 above)
Comment (JosephUoA) [reply stac_generator/raster_data/generator.py:22]: I think its fine to have the mapping dictionary prepared for the future.

File: stac_generator/raster_data/generator.py [hunk 23]
```diff
@@ -95,10 +116,16 @@ def create_item_from_config(self, source_cfg: RasterSourceConfig) -> list[pystac
...
                 if source_cfg.datetime.tzinfo is None:
-                    datetime_aware = source_cfg.datetime.replace(tzinfo=UTC)
+                    datetime_aware = source_cfg.datetime.replace(tzinfo=file_timezone)
                 else:
                     datetime_aware = source_cfg.datetime
 
+                epsg = crs.to_epsg()
+                if source_cfg.epsg != epsg:
+                    raise ValueError(
+                        f"EPSG mismatch: source_cfg.epsg ({source_cfg.epsg}) does not match shapefile EPSG ({epsg})."
```
Comment (JosephUoA) [inline stac_generator/raster_data/generator.py]: This is not a shapefile.

File: example/csv/generated/raster/orthomosaic_item/orthomosaic_item.json [hunk 24]
```diff
@@ -41,8 +41,6 @@
     6951062.348691959
   ],
   "properties": {
-    "eo:snow_cover": 0,
-    "eo:cloud_cover": 0,
```
Comment (JosephUoA) [inline example/csv/generated/raster/orthomosaic_item/orthomosaic_item.json]: Not sure what code changes caused these two lines to go away but I think they should still be here.

Comment (alisha17) [reply example/csv/generated/raster/orthomosaic_item/orthomosaic_item.json]: it went away because I used .apply() for eo extension but have fixed it now

File: stac_generator/raster_data/generator.py (hunk 23 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: fixed

**Comment by JosephUoA** [comment]
approved once CI is fixed

File: stac_generator/raster_data/schema.py (hunk 21 above)
Comment (alisha17) [reply stac_generator/raster_data/schema.py]: removed

File: stac_generator/raster_data/generator.py (hunk 18 above)
Comment (alisha17) [reply stac_generator/raster_data/generator.py]: removed
